quantum/debounce/sym_pk.c

```diff
@@ -0,0 +1,120 @@
+
+static uint16_t custom_wrap_timer_reference  = 0;
+static uint8_t  custom_wrap_timer_last_value = 0;
+
+static uint8_t custom_wrap_timer_read(void) {
+    uint16_t custom_wrap_timer_new_reference = timer_read();
+    uint16_t diff                            = custom_wrap_timer_new_reference - custom_wrap_timer_reference;
+    custom_wrap_timer_reference              = custom_wrap_timer_new_reference;
+    custom_wrap_timer_last_value             = (custom_wrap_timer_last_value + diff) % (MAX_DEBOUNCE + 1);
+    return custom_wrap_timer_last_value;
+}
+
+void update_debounce_counters(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time);
+void transfer_matrix_values(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time);
+
+// we use num_rows rather than MATRIX_ROWS to support split keyboards
+void debounce_init(uint8_t num_rows) {
+    debounce_counters = (debounce_counter_t *)malloc(num_rows * MATRIX_COLS * sizeof(debounce_counter_t));
+    int i             = 0;
+    for (uint8_t r = 0; r < num_rows; r++) {
+        for (uint8_t c = 0; c < MATRIX_COLS; c++) {
+            debounce_counters[i++] = DEBOUNCE_ELAPSED;
+        }
+    }
+}
+
+void debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
+    uint8_t current_time = custom_wrap_timer_read();
+    if (counters_need_update) {
+        update_debounce_counters(raw, cooked, num_rows, current_time);
+    }
+
+    if (changed) {
+        transfer_matrix_values(raw, cooked, num_rows, current_time);
+    }
+}
+
+// If the current time is > debounce counter, set the counter to enable input.
+void update_debounce_counters(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time) {
+    counters_need_update                 = false;
+    debounce_counter_t *debounce_pointer = debounce_counters;
+    for (uint8_t row = 0; row < num_rows; row++) {
+        for (uint8_t col = 0; col < MATRIX_COLS; col++) {
+            if (*debounce_pointer != DEBOUNCE_ELAPSED) {
+                if (TIMER_DIFF(current_time, *debounce_pointer, MAX_DEBOUNCE) >= DEBOUNCE) {
+                    *debounce_pointer = DEBOUNCE_ELAPSED;
+                    cooked[row]       = (cooked[row] & ~(ROW_SHIFTER << col)) | (raw[row] & (ROW_SHIFTER << col));
+                } else {
+                    counters_need_update = true;
+                }
+            }
+            debounce_pointer++;
+        }
+    }
+}
+
+// upload from raw_matrix to final matrix;
+void transfer_matrix_values(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time) {
+    debounce_counter_t *debounce_pointer = debounce_counters;
+    for (uint8_t row = 0; row < num_rows; row++) {
+        matrix_row_t delta = raw[row] ^ cooked[row];
+        for (uint8_t col = 0; col < MATRIX_COLS; col++) {
+            if (delta & (ROW_SHIFTER << col)) {
+                if (*debounce_pointer == DEBOUNCE_ELAPSED) {
```

Comment:
This might be a potential game-breaker?
This check (*debounce_pointer == DEBOUNCE_ELAPSED) would prevent extending the timer.
In traditional `sym_g`, the timer gets reset if there is any change. Here you only reset the timer if it is not running && there is a change.

This means if you press a key down, you simply wait 5ms and take the final input, instead of waiting for a 5ms period of stabilization. Here's a diagram of how i think this works as the code is:
This implementation (ignores everything between raw change and MERemain=0)
The raw input where it says X are completely ignored
```
Raw input:   | 1 | X | X | X | X | 1 | 0 | X | X | X | X | 0 |
Sent input:  | 0 | 0 | 0 | 0 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 0 |
MSRemain:    | 5 | 4 | 3 | 2 | 1 | 0 | 5 | 4 | 3 | 2 | 1 | 0 |
MSTime:      | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11| 12|
```

Reference implementation (sym_g)
```
Raw input:   | 1 | 0 | 0 | 0 | 1 | 1 | 0 | 0 | 0 | 0 | 0 | 0 |
Sent input:  | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |
MSRemain:    | 5 | 5 | 4 | 3 | 5 | 4 | 5 | 4 | 3 | 2 | 1 | 0 |
MSTime:      | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11| 12|
```

AFAICT with this single line, your algorithm appears to be eager_pk with a send delay rather than 
sym_pk? Feel free to correct me if i have misunderstood it :)

Reply:
I think you are misunderstanding my code.
I am not tracking if there was any change in the signal. I am tracking if there is any difference between the Raw and the Cooked value.
If there is a difference (Delta = 1), and the counter is not started, then I start the counter. This only happens on:
* 1->0 transitions on the Raw signal, when the Cooked signal is 1.
* 0->1 transitions on the Raw signal, when the Cooked signal is 0.

If there is no difference then there is no point in keeping the counter running, because there's nothing to push at the end of the Debounce time period, because the cooked value is already what we would have pushed, so I just reset it to Elapsed.
This happens if there is:
* 1->0 transitions on the Raw signal, when the Cooked signal is 0.
* 0->1 transitions on the Raw signal, when the Cooked signal is 1.
* If there are no transitions, but the Raw signal equals the Cooked signal. -- this has no effect, because the timer is already elapsed. But it's more efficient this way, rather then looking for per-bit transition on the raw line. I don't have to store yet another matrix this way.

So if there's a change that tries to induce a change in the output then I restart the timer, running.
If there's a change that goes in the direction in which we already are, then I just reset the timer to idle. You don't have to count time on both 1->0 and 0->1 changes. Only one will suffice, depending on what your cooked value is.

Please see this hand-drawn waveform:
```
Current Time: | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11| 12| 13| 14|
Raw Input:    | 0 | 1 | 1 | 1 | 0 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 |
Cooked Input: | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 1 |
Delta:        | 0 | 1 | 1 | 1 | 0 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 0 | 0 |
Counter:      | EL| 2 | 2 | 2 | EL| EL| 7 | 7 | 7 | 7 | 7 | 7 | EL| EL|
DIFF:         |N/A| 0 | 1 | 2 |N/A|N/A| 0 | 1 | 2 | 3 | 4 | 5 |N/A|N/A|
```

If you disagree, please re-open this conversation.

Reply:
**I see the timer reset now; you're right, no storage is required.**
 
> If there is no difference then there is no point in keeping the counter running, because there's nothing to push at the end of the Debounce time period, because the cooked value is already what we would have pushed, so I just reset it to Elapsed.

Right, this is my point of contention; i couldn't see a reset to Elapsed when you had an active timer. I see that now, On line 111

> 
> Please see this hand-drawn waveform:
> 
> ```
> Current Time: | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11| 12| 13| 14|
> Raw Input:    | 0 | 1 | 1 | 1 | 0 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 |
> Cooked Input: | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 1 |
> Delta:        | 0 | 1 | 1 | 1 | 0 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 0 | 0 |
> Counter:      | EL| 2 | 2 | 2 | EL| EL| 7 | 7 | 7 | 7 | 7 | 7 | EL| EL|
> DIFF:         |N/A| 0 | 1 | 2 |N/A|N/A| 0 | 1 | 2 | 3 | 4 | 5 |N/A|N/A|
> ```
> 

Excellent diagram. This is what i assumed you were going for, and the code does actually reflect that. I incorrectly assumed that the code worked like my previous diagram due to my misjudgement :)